src/core/ext/xds/xds_bootstrap.cc

```diff
@@ -98,39 +100,118 @@ grpc_error_handle ParseChannelCredsArray(const Json::Array& json,
+        grpc_insecure_credentials_create());
+  }
+  bool IsValidConfig(const Json& /*config*/) const override { return true; }
+  const char* creds_type() const override { return "insecure"; }
+};
+
+class FakeXdsChannelCredsFactory : public XdsChannelCredsFactory {
+ public:
+  RefCountedPtr<grpc_channel_credentials> CreateXdsChannelCreds(
+      const Json& /*config*/) const override {
+    return RefCountedPtr<grpc_channel_credentials>(
+        grpc_fake_transport_security_credentials_create());
+  }
+  bool IsValidConfig(const Json& /*config*/) const override { return true; }
+  const char* creds_type() const override { return "fake"; }
+};
+
 //
 // XdsChannelCredsRegistry
 //
 
+XdsChannelCredsRegistryState::XdsChannelCredsRegistryState() {
+  RegisterXdsChannelCredsFactory(
+      absl::make_unique<GoogleDefaultXdsChannelCredsFactory>());
+  RegisterXdsChannelCredsFactory(
+      absl::make_unique<InsecureXdsChannelCredsFactory>());
+  RegisterXdsChannelCredsFactory(
+      absl::make_unique<FakeXdsChannelCredsFactory>());
+}
+
 bool XdsChannelCredsRegistry::IsSupported(const std::string& creds_type) {
-  return creds_type == "google_default" || creds_type == "insecure" ||
-         creds_type == "fake";
+  return g_state->LookupXdsChannelCredsFactory(creds_type) != nullptr;
 }
 
-bool XdsChannelCredsRegistry::IsValidConfig(const std::string& /*creds_type*/,
-                                            const Json& /*config*/) {
-  // Currently, none of the creds types actually take a config, but we
-  // ignore whatever might be specified in the bootstrap file for
-  // forward compatibility reasons.
-  return true;
+bool XdsChannelCredsRegistry::IsValidConfig(const std::string& creds_type,
+                                            const Json& config) {
+  XdsChannelCredsFactory* factory =
+      g_state->LookupXdsChannelCredsFactory(creds_type);
+  return factory != nullptr && factory->IsValidConfig(config);
 }
 
 RefCountedPtr<grpc_channel_credentials>
 XdsChannelCredsRegistry::MakeChannelCreds(const std::string& creds_type,
-                                          const Json& /*config*/) {
-  if (creds_type == "google_default") {
-    return RefCountedPtr<grpc_channel_credentials>(
-        grpc_google_default_credentials_create(nullptr));
-  } else if (creds_type == "insecure") {
-    return RefCountedPtr<grpc_channel_credentials>(
-        grpc_insecure_credentials_create());
-  } else if (creds_type == "fake") {
-    return RefCountedPtr<grpc_channel_credentials>(
-        grpc_fake_transport_security_credentials_create());
-  }
-  return nullptr;
+                                          const Json& config) {
+  XdsChannelCredsFactory* factory =
+      g_state->LookupXdsChannelCredsFactory(creds_type);
+  if (factory == nullptr) return nullptr;
+  return factory->CreateXdsChannelCreds(config);
+}
+
+void XdsChannelCredsRegistry::Init() {
+  if (g_state == nullptr) g_state = new XdsChannelCredsRegistryState();
```

Comment:
I think we should just unconditionally set it here, without a check to see that it's null.